### PR TITLE
fix(permissions): reject malformed doap and ap create/update request (DSP-1328)

### PR DESCRIPTION
**Pull request by SepidehAlassi**
resolves DSP-1328

- [x] Improve documentation about permissions
- [x] Reject malformed update `hasPermissions` requests for both AP and DOAP
- [x] Reject create AP and DOAP requests when `hasPermissions` was malformed.
- [x] add tests

### CODE REVIEW

File: docs/03-apis/api-admin/permissions.md [hunk 1]
```diff
@@ -79,11 +79,40 @@ As a response, the created administrative permission and its IRI are returned as
     }
 }
 ```
+`hasPermissions` contains permission types that must be granted. See [the complete description of administrative
+permission types](../../05-internals/design/api-admin/administration.md#administrative-permissions).
+In summary, each permission should contain followings:
+- `name` : indicates the type of the type of the permission that can be one of the followings:
```
Comment (subotic) [inline docs/03-apis/api-admin/permissions.md:85]: probably "the type of the type of permission" -> "the type of permission"

File: docs/03-apis/api-admin/permissions.md [hunk 2]
```diff
@@ -101,6 +130,25 @@ default object access permission for a group of a project the request body would
...
+permission types](../../05-internals/design/api-admin/administration.md#default-object-access-permissions). 
+In summary, each permission should contain followings:
+- `additionalInformation`: To whom the permission should be granted: project members, known users, unknown users, etc.
+- `name` : indicates the type of the permission that can be one of the followings.
+   - `RV`: restricted view permission (least privileged)
+   - `V`: view permission
+   - `M` modify permission
+   - `D`: delete permission
+   - `CR`: change rights permission (most privileged)
+- `permissionCode`: The code assigned to a permission indicating its hierarchical lever. These codes are as below:
```
Comment (subotic) [inline docs/03-apis/api-admin/permissions.md:143]: "lever" -> "level"?

File: webapi/src/main/scala/org/knora/webapi/messages/admin/responder/permissionsmessages/PermissionsMessagesUtilADM.scala [hunk 3]
```diff
@@ -88,4 +105,82 @@ object PermissionsMessagesUtilADM {
 
     CacheUtil.remove(PermissionsCacheName, key)
   }
+
```
Comment (subotic) [inline webapi/src/main/scala/org/knora/webapi/messages/admin/responder/permissionsmessages/PermissionsMessagesUtilADM.scala:108]: Missing docs

File: webapi/src/main/scala/org/knora/webapi/messages/admin/responder/permissionsmessages/PermissionsMessagesUtilADM.scala [hunk 4]
```diff
@@ -88,4 +105,82 @@ object PermissionsMessagesUtilADM {
...
+        additionalInformation = None,
+        permissionCode = None
+      )
+    }
+    updatedPermissions
+  }
+
+  /**
+    * For default object access permission, we need to make sure that the value given for the projectCode matches the value of name parameter.
+    * This method, validates the content of hasPermissions collection by verifying that both projectCode and name indicate the same type of permission.
```
Comment (subotic) [inline webapi/src/main/scala/org/knora/webapi/messages/admin/responder/permissionsmessages/PermissionsMessagesUtilADM.scala:162]: what is meant by projectCode?

**Review by subotic — APPROVED** [review]
LGTM, thanks!

File: webapi/src/main/scala/org/knora/webapi/messages/admin/responder/permissionsmessages/PermissionsMessagesUtilADM.scala (hunk 4 above)
Comment (SepidehAlassi) [reply webapi/src/main/scala/org/knora/webapi/messages/admin/responder/permissionsmessages/PermissionsMessagesUtilADM.scala:162]: hehe that was a permissioncode, sorry for the typo.

**Pull request merged**